File: frontend/src/data/seed.ts
import type { MissionCadence, Product, ProductFamily, Profile } from '../types'
import { uid } from '../lib/db'
import { DEFAULT_CADENCES } from '../lib/settings'

const ENTITIES = [
  { code: 'AFKOT' as const, name: 'AFKOT' },
  { code: 'BOSCAM' as const, name: 'BOSCAM' },
  { code: 'CTC' as const, name: 'CTC' },
]

const ZONES = ['ZONNE EST-SUD', 'ZONNE LITTORAL-OUEST', 'ZONNE GRAND NORD']

const SITES: { zone: string; name: string }[] = [
  { zone: 'ZONNE EST-SUD', name: 'GAROUA BOULAI' },
  { zone: 'ZONNE EST-SUD', name: 'BERTOUA' },
  { zone: 'ZONNE LITTORAL-OUEST', name: 'DOUALA' },
  { zone: 'ZONNE LITTORAL-OUEST', name: 'KUMBA' },
  { zone: 'ZONNE LITTORAL-OUEST', name: 'BAMENDA' },
  { zone: 'ZONNE LITTORAL-OUEST', name: 'BAFOUSSAM' },
  { zone: 'ZONNE GRAND NORD', name: 'MAROUA' },
  { zone: 'ZONNE GRAND NORD', name: 'GAROUA' },
  { zone: 'ZONNE GRAND NORD', name: 'NGAOUNDERE' },
]

export async function buildSeedData() {
  const zones = ZONES.map((name) => ({ id: uid('ZON-'), name, active: true }))
  const zoneByName = Object.fromEntries(zones.map((z) => [z.name, z.id]))

  const sites = SITES.map((s) => ({
    id: uid('SITE-'),
    zoneId: zoneByName[s.zone],
    name: s.name,
    active: true,
  }))

  const stores = sites.flatMap((site) =>
    ENTITIES.map((e) => ({
      id: uid('MAG-'),
      siteId: site.id,
      name: `${e.code} — ${sites.find((x) => x.id === site.id)!.name}`,
      odooLocationCode: `${e.code}/${sites.find((x) => x.id === site.id)!.name}`,
      active: true,
    }))
  )

  const entities = ENTITIES.map((e) => ({
    id: uid('ENT-'),
    code: e.code,
    name: e.name,
    active: true,
  }))

  const entityByCode = Object.fromEntries(entities.map((e) => [e.code, e.id]))
  const products: Product[] = []

  const afkotCig = [
    'Esse change 4mg', 'Esse change plus', 'Time change simple', 'Time change plus',
    'Time apple V', 'Time tropic V', 'Time gold V', 'Time red', 'Time strawberry V',
  ]
  afkotCig.forEach((d, i) =>
    products.push({
      id: uid('PRD-'),
      entityId: entityByCode.AFKOT,
      family: 'CIGARETTE' as ProductFamily,
      code: `AFK-C${String(i + 1).padStart(3, '0')}`,
      designation: d,
      brand: d.startsWith('Esse') ? 'Esse' : 'Time',
      packaging: 'Carton/Cartouche',
      unitPrice: 90,
      packPrice: 4500,
      active: true,
    })
  )

  const afkotGadget = [
    'Affiches Time', 'Affiches A3', 'Caisses', 'Casquettes', 'Presentoirs',
    'Parassols', 'Polos', 'Sacs plastiques',
  ]
  afkotGadget.forEach((d, i) =>
    products.push({
      id: uid('PRD-'),
      entityId: entityByCode.AFKOT,
      family: 'GADGET' as ProductFamily,
      code: `AFK-G${String(i + 1).padStart(3, '0')}`,
      designation: d,
      brand: 'Time',
      packaging: 'Balot/Unité',
      unitPrice: 500,
      packPrice: 50000,
      active: true,
    })
  )

  const bosVapes = [
    'vega str watermelon 0,8p', 'vega blueberry ice 0,8p', 'vega banana ice 0,8p',
    'vega str ice 0,8p', 'vega menthol ice 0,8p', 'vega str banana 1,5p',
    'vega grape-ICE 1,5p', 'vega str ice 1,5p',
  ]
  bosVapes.forEach((d, i) =>
    products.push({
      id: uid('PRD-'),
      entityId: entityByCode.BOSCAM,
      family: 'VAPE' as ProductFamily,
      code: `BOS-V${String(i + 1).padStart(3, '0')}`,
      designation: d,
      brand: 'Vega',
      packaging: 'Carton/Paquet/Pièce',
      unitPrice: 3500,
      packPrice: 350000,
      active: true,
    })
  )

  const bosCig = ['vega b10 v', 'vega b10 sv', 'vega b20sv', 'vega m10 v', 'vega ks apple mint 20 v', 'vega ks ice 20 v']
  bosCig.forEach((d, i) =>
    products.push({
      id: uid('PRD-'),
      entityId: entityByCode.BOSCAM,
      family: 'CIGARETTE' as ProductFamily,
      code: `BOS-C${String(i + 1).padStart(3, '0')}`,
      designation: d,
      brand: 'Vega',
      packaging: 'Carton/Cartouche',
      unitPrice: 120,
      packPrice: 6000,
      active: true,
    })
  )

  const ctcCig = ['GOLD SEAL B10 SV', 'GOLD SEAL B20 V', 'D&J M20 SV', 'ORIS PULSE YELLOW FIZZ SS 20']
  ctcCig.forEach((d, i) =>
    products.push({
      id: uid('PRD-'),
      entityId: entityByCode.CTC,
      family: 'CIGARETTE' as ProductFamily,
      code: `CTC-C${String(i + 1).padStart(3, '0')}`,
      designation: d,
      brand: d.includes('GOLD') ? 'Gold Seal' : d.includes('ORIS') ? 'Oris' : 'D&J',
      packaging: 'Carton/Cartouche',
      unitPrice: 100,
      packPrice: 5000,
      active: true,
    })
  )

  const profiles: Profile[] = [
    {
      id: uid('USR-'),
      name: 'Contrôleur Terrain',
      role: 'CONTROLEUR',
      entityIds: entities.map((e) => e.id),
      zoneIds: zones.map((z) => z.id),
      passwordHash: '',
      active: true,
    },
    {
      id: uid('USR-'),
      name: 'Direction CDG ATG',
      role: 'DIRECTION',
      entityIds: entities.map((e) => e.id),
      zoneIds: zones.map((z) => z.id),
      passwordHash: '',
      active: true,
    },
    {
      id: uid('USR-'),
      name: 'Administrateur ICAMS',
      role: 'ADMIN',
      entityIds: entities.map((e) => e.id),
      zoneIds: zones.map((z) => z.id),
      passwordHash: '',
      active: true,
    },
  ]

  const cadences: MissionCadence[] = DEFAULT_CADENCES.map((c) => ({
    id: uid('CAD-'),
    ...c,
    active: true,
  }))

  return { zones, sites, stores, entities, products, profiles, cadences }
}

export function familyLabel(f: ProductFamily): string {
  if (f === 'CIGARETTE') return 'Cigarettes'
  if (f === 'GADGET') return 'Gadgets'
  return 'Vapes'
}

export const FAMILIES_BY_ENTITY = {
  AFKOT: ['CIGARETTE', 'GADGET'] as ProductFamily[],
  BOSCAM: ['CIGARETTE', 'GADGET', 'VAPE'] as ProductFamily[],
  CTC: ['CIGARETTE', 'GADGET'] as ProductFamily[],
}

File: backend/app/seed.py
import json
import uuid
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

from app.models import (
    AppSettingsRow,
    Entity,
    InventoryLine,
    Mission,
    MissionCadence,
    OdooSnapshot,
    Product,
    Profile,
    Site,
    Store,
    SyncQueueItem,
    Zone,
)
from app.schemas import (
    AppSettingsSchema,
    ConversionRules,
    OdooColumnMapping,
    ThemeSettings,
)

DEFAULT_SETTINGS = AppSettingsSchema(
    companyName="ATG — Inventory Control & Audit",
    currency="FCFA",
    theme=ThemeSettings(
        primaryColor="#4f46e5",
        accentColor="#818cf8",
        backgroundColor="#020617",
    ),
    conversionRules=ConversionRules(
        cigaretteCartonToCartouches=50,
        gadgetBalotToUnits=100,
        vapeCartonToPieces=100,
        vapePaquetToPieces=10,
    ),
    odooImportMapping=OdooColumnMapping(
        productCode="default_code",
        productName="name",
        location="location",
        quantity="quantity",
        uom="uom",
    ),
    odooExportPrefix="ICAMS_AJUSTEMENT",
    observationTemplates=[
        "RAS",
        "manquant constaté",
        "excédent constaté",
        "à régulariser",
        "confusion de parfum",
        "confusion nouvelle/ancienne image",
        "écart non livré",
    ],
    missionStatuses=["BROUILLON", "EN_COURS", "SOUMIS", "VALIDE", "EXPORTE"],
    syncRetryMinutes=5,
    allowOfflineMode=True,
    requireOdooSnapshotBeforeCount=False,
)

DEFAULT_CADENCES = [
    {"name": "Journalier", "intervalDays": 1, "description": "Contrôle quotidien", "sortOrder": 1},
    {"name": "Mensuel", "intervalDays": 30, "description": "Inventaire mensuel", "sortOrder": 2},
    {"name": "Bimensuel", "intervalDays": 60, "description": "Tous les deux mois", "sortOrder": 3},
    {"name": "Trimestriel", "intervalDays": 90, "description": "Inventaire trimestriel", "sortOrder": 4},
    {"name": "Annuel", "intervalDays": 365, "description": "Inventaire annuel", "sortOrder": 5},
]

ENTITIES = [
    {"code": "AFKOT", "name": "AFKOT"},
    {"code": "BOSCAM", "name": "BOSCAM"},
    {"code": "CTC", "name": "CTC"},
]

ZONES = ["ZONNE EST-SUD", "ZONNE LITTORAL-OUEST", "ZONNE GRAND NORD"]

SITES = [
    {"zone": "ZONNE EST-SUD", "name": "GAROUA BOULAI"},
    {"zone": "ZONNE EST-SUD", "name": "BERTOUA"},
    {"zone": "ZONNE LITTORAL-OUEST", "name": "DOUALA"},
    {"zone": "ZONNE LITTORAL-OUEST", "name": "KUMBA"},
    {"zone": "ZONNE LITTORAL-OUEST", "name": "BAMENDA"},
    {"zone": "ZONNE LITTORAL-OUEST", "name": "BAFOUSSAM"},
    {"zone": "ZONNE GRAND NORD", "name": "MAROUA"},
    {"zone": "ZONNE GRAND NORD", "name": "GAROUA"},
    {"zone": "ZONNE GRAND NORD", "name": "NGAOUNDERE"},
]


def uid(prefix: str = "") -> str:
    return f"{prefix}{uuid.uuid4().hex[:8].upper()}"


def seed_database(db: Session) -> None:
    if db.query(Zone).count() > 0:
        return

    zones = [Zone(id=uid("ZON-"), name=name, active=True) for name in ZONES]
    zone_by_name = {z.name: z.id for z in zones}
    db.add_all(zones)

    sites = [
        Site(id=uid("SITE-"), zone_id=zone_by_name[s["zone"]], name=s["name"], active=True)
        for s in SITES
    ]
    db.add_all(sites)
    site_by_id = {s.id: s for s in sites}

    stores: list[Store] = []
    for site in sites:
        for ent in ENTITIES:
            stores.append(
                Store(
                    id=uid("MAG-"),
                    site_id=site.id,
                    name=f"{ent['code']} — {site.name}",
                    odoo_location_code=f"{ent['code']}/{site.name}",
                    active=True,
                )
            )
    db.add_all(stores)

    entities = [Entity(id=uid("ENT-"), code=e["code"], name=e["name"], active=True) for e in ENTITIES]
    db.add_all(entities)
    entity_by_code = {e.code: e.id for e in entities}

    products: list[Product] = []

    afkot_cig = [
        "Esse change 4mg", "Esse change plus", "Time change simple", "Time change plus",
        "Time apple V", "Time tropic V", "Time gold V", "Time red", "Time strawberry V",
    ]
    for i, d in enumerate(afkot_cig):
        products.append(
            Product(
                id=uid("PRD-"),
                entity_id=entity_by_code["AFKOT"],
                family="CIGARETTE",
                code=f"AFK-C{i + 1:03d}",
                designation=d,
                brand="Esse" if d.startswith("Esse") else "Time",
                packaging="Carton/Cartouche",
                unit_price=90,
                pack_price=4500,
                active=True,
            )
        )

    afkot_gadget = [
        "Affiches Time", "Affiches A3", "Caisses", "Casquettes", "Presentoirs",
        "Parassols", "Polos", "Sacs plastiques",
    ]
    for i, d in enumerate(afkot_gadget):
        products.append(
            Product(
                id=uid("PRD-"),
                entity_id=entity_by_code["AFKOT"],
                family="GADGET",
                code=f"AFK-G{i + 1:03d}",
                designation=d,
                brand="Time",
                packaging="Balot/Unité",
                unit_price=500,
                pack_price=50000,
                active=True,
            )
        )

    bos_vapes = [
        "vega str watermelon 0,8p", "vega blueberry ice 0,8p", "vega banana ice 0,8p",
        "vega str ice 0,8p", "vega menthol ice 0,8p", "vega str banana 1,5p",
        "vega grape-ICE 1,5p", "vega str ice 1,5p",
    ]
    for i, d in enumerate(bos_vapes):
        products.append(
            Product(
                id=uid("PRD-"),
                entity_id=entity_by_code["BOSCAM"],
                family="VAPE",
                code=f"BOS-V{i + 1:03d}",
                designation=d,
                brand="Vega",
                packaging="Carton/Paquet/Pièce",
                unit_price=3500,
                pack_price=350000,
                active=True,
            )
        )

    bos_cig = ["vega b10 v", "vega b10 sv", "vega b20sv", "vega m10 v", "vega ks apple mint 20 v", "vega ks ice 20 v"]
    for i, d in enumerate(bos_cig):
        products.append(
            Product(
                id=uid("PRD-"),
                entity_id=entity_by_code["BOSCAM"],
                family="CIGARETTE",
                code=f"BOS-C{i + 1:03d}",
                designation=d,
                brand="Vega",
                packaging="Carton/Cartouche",
                unit_price=120,
                pack_price=6000,
                active=True,
            )
        )

    ctc_cig = ["GOLD SEAL B10 SV", "GOLD SEAL B20 V", "D&J M20 SV", "ORIS PULSE YELLOW FIZZ SS 20"]
    for i, d in enumerate(ctc_cig):
        brand = "Gold Seal" if "GOLD" in d else "Oris" if "ORIS" in d else "D&J"
        products.append(
            Product(
                id=uid("PRD-"),
                entity_id=entity_by_code["CTC"],
                family="CIGARETTE",
                code=f"CTC-C{i + 1:03d}",
                designation=d,
                brand=brand,
                packaging="Carton/Cartouche",
                unit_price=100,
                pack_price=5000,
                active=True,
            )
        )

    db.add_all(products)

    all_entity_ids = [e.id for e in entities]
    all_zone_ids = [z.id for z in zones]
    profiles = [
        Profile(
            id=uid("USR-"),
            name="Contrôleur Terrain",
            role="CONTROLEUR",
            entity_ids=json.dumps(all_entity_ids),
            zone_ids=json.dumps(all_zone_ids),
            password_hash="",
            active=True,
        ),
        Profile(
            id=uid("USR-"),
            name="Direction CDG ATG",
            role="DIRECTION",
            entity_ids=json.dumps(all_entity_ids),
            zone_ids=json.dumps(all_zone_ids),
            password_hash="",
            active=True,
        ),
        Profile(
            id=uid("USR-"),
            name="Administrateur ICAMS",
            role="ADMIN",
            entity_ids=json.dumps(all_entity_ids),
            zone_ids=json.dumps(all_zone_ids),
            password_hash="",
            active=True,
        ),
    ]
    db.add_all(profiles)

    cadences = [
        MissionCadence(
            id=uid("CAD-"),
            name=c["name"],
            interval_days=c["intervalDays"],
            description=c.get("description"),
            active=True,
            sort_order=c["sortOrder"],
        )
        for c in DEFAULT_CADENCES
    ]
    db.add_all(cadences)

    db.add(AppSettingsRow(id="global", data=DEFAULT_SETTINGS.model_dump_json()))
    db.commit()
